docs/generic-constraints.md

```diff
@@ -0,0 +1,142 @@
+> Ideally, this behaviour should work out of the box, but, could be solved with updates to the type engine.
+
+## Design
+### Syntax
+There are two new options for syntax that come with this generic constraints.
+1. Treat the generic parameter as a variable--annotate it as declaring a new variable.
+```luau
+local function getProperty<T, K: keyof<T>>( object: T, key: K ): index<T, K>
+	return object[ key ]
+end
+
+type getProperty<T, K: keyof<T>> = ( object: T, key: K ) -> index<T, K>
+```
+This would match the current expectation of `:` being used to type annotate a variable. This would be backwards compatible without major performance implications. As a note, the ordering of `T` and `K` is arbitrary and can be switched if desired.
+
+2. Use a `where` clause.
+```luau
+local function getProperty<T, K where K: keyof<T>>( object: T, key: K ): index<T, K>
+	return object[ key ]
+end
+
+type getProperty<T, K where K: keyof<T>> = ( object: T, key: K ) -> ( index<T, K> )
+```
+This would allow users to specify the constraints of the separately from the generic declaration itself. This would be reminiscent to users of C#. Imposing multiple constraints on different generics can be done by delimiting with `,`. This would be backwards compatible, without major performance implications, as it could just be a conditional keyword. 
+
+The `where` clause syntax can be further expanded to pose custom bounds on generics.
+```luau
+local function add<T, K where add<T, K>>( a: T, b: K )
+	return a + b
+end
+```
+The `add` function would only accept `T` and `K` if `add<T, K` is not `never`. This is currently available in Luau, but this behaviour is locked behind `add<T, K>` being the return type.
+
+### Usage
+#### Subtyping Constraints
+This is the problem posed at the beginning of this RFC.
+```lua
+local function callbackProperty<T, K: keyof<T>>( object: T, key: K, callback: (property: index<T, K>) -> () )
+	callback( object[ key ] )
+end
+```
+`K` will be constrained to be `keyof<T>`. This would allow `index<T, K>` to infer the type properly.
+#### Equality Constraints
+Suppose we wanted to make a function that could only add `number`s or `vector`s.
+```luau
+-- Attempt 1
+local function sumSpecific( a: number | vector, b: number | vector ): number | vector
+	return a + b -- not okay!
+end
+-- Attempt 2
+local function sumSpecific<T>( a: T, b: T ): T
+	return a + b -- not okay! not all types are addable.
+end
+```
+This is to be expected. Luau does not interpret `a` and `b` as the same type, but if we use a generic then Luau throws an error that not all types are addable.
+```luau
+-- Attempt 3
+local sumSpecific: ( ( number, number ) -> ( number ) ) & ( ( vector, vector ) -> ( vector ) ) = function( a, b )
+	return a + b -- okay!
+end
+```
+This is fine, but is not concise whatsoever, and does not allow for the traditional function declaration syntax.
+```luau
+local function sumSpecific<T: number | vector>( a: T, b: T ): T
+	return a + b
+end
```

Comment:
It's a bit hard to do this. A counterexample:

```luau
function f() -- f : () -> number | string
  return if math.random() > 0.5 then 5 else vector.create(0, 0, 0)
end

sumSpecific(f(), f())
```

Which would still infer `T` to be `number | vector`, and provide a `number` for `a` and `vector` for `b`.